File: cpp/algos_ds/graph/graph.cpp
#include <iostream>
#include <vector>
#include <utility>
#include <sstream>
#include <queue>
#include <stack>
using namespace std;

struct Edge {
    int prev;
    int next;
};

class Graph {
    int num_nodes;
    vector<vector<Edge>> adjacency_list;
public: 
    Graph(int n) : num_nodes(n) {
        adjacency_list.resize(n, vector<Edge>());
    }
    ~Graph() {
    }
    void add_edge(int prev, int next) {
        adjacency_list[prev].push_back(Edge{prev, next});
        adjacency_list[next].push_back(Edge{next, prev});
    }
    string bfs_traversal(int start) {
        stringstream path;
        queue<pair<int, int>> q;
        vector<bool> visited;
        visited.resize(adjacency_list.size(), false);
        q.push(pair<int, int> (start, 1));
        while(!q.empty()) {
            pair<int, int> current = q.front();
            q.pop();
            if(!visited[current.first]) {
                visited[current.first] = true;
                path << "->(" << current.first << ", " << current.second << ")";
                for(Edge &e : adjacency_list[current.first]) {
                    q.push(pair<int, int>(e.next, current.second + 1));
                }
            }
        }
        return path.str();
    }
    string dfs_traversal(int start) {
        stringstream path;
        stack<pair<int, int>> s;
        vector<bool> visited;
        visited.resize(adjacency_list.size(), false);
        s.push(pair<int, int> (start, 1));
        while(!s.empty()) {
            pair<int, int> current = s.top();
            s.pop();
            if(!visited[current.first]) {
                visited[current.first] = true;
                path << "->(" << current.first << ", " << current.second << ")";
                for(Edge &e : adjacency_list[current.first]) {
                    s.push(pair<int, int>(e.next, current.second + 1));
                }
            }
        }
        return path.str();
    }
    operator string() {
        stringstream ss;
        for(int i = 0; i < num_nodes; i++) {
            ss << i << "-";
            for(Edge &j : adjacency_list[i]) {
                ss << "->" << j.next;
            }
            ss << endl;
        }
        return ss.str();
    }
};

ostream& operator<<(ostream &stream, Graph &graph) {
    stream << string(graph);
    return stream;
}

int main() {
    int num_graphs;
    cin >> num_graphs;
    while(num_graphs--) {
        int nodes, edges;
        scanf("%d %d", &nodes, &edges);
        Graph graph = Graph(nodes);
        while(edges--) {
            int prev, next;
            scanf("%d-%d", &prev, &next);
            graph.add_edge(prev, next);
        }
        vector<int> tests = {0, 6, 3, 7};
        cout << "-- Graph --\n" << graph << endl;
        for(int i : tests) {
            cout << "-- BFS Traversal --\n" << graph.bfs_traversal(i) << endl;
        }
        for(int i : tests) {
            cout << "-- DFS Traversal --\n" << graph.dfs_traversal(i) << endl;
        }
    }
}

File: python/algos_ds/graph/graph.py
import queue

class Edge:
    def __init__(self, p, n, c=0):
        self.previous = p
        self.next = n
        self.cost = c

class Graph:
    def __init__(self, size):
        self.num_nodes = size
        self.adjacency_list = [list() for x in range(size)]

    def add_edge(self, p, n, c=0):
        self.adjacency_list[p].append(Edge(p, n, c))
        self.adjacency_list[n].append(Edge(n, p, c))

    def __str__(self):
        string = list()
        for i, lst in enumerate(self.adjacency_list):
            string.append(f'{i}-')
            for edge in lst:
                string.append(f'->{edge.next}')
            string.append('\n')
        return ''.join(map(str, string))

    def bfs_traversal(self, start):
        string = list()
        q = queue.Queue()
        visited = [False for x in range(self.num_nodes)]
        q.put((start, 1))
        while not q.empty():
            node, level = q.get()
            if not visited[node]:
                visited[node] = True
                string.append(f'->({node}, {level})')
                for edge in self.adjacency_list[node]:
                    q.put((edge.next, level + 1))
        return ''.join(map(str, string))

    def dfs_traversal(self, start):
        string = list()
        q = queue.LifoQueue()
        visited = [False for x in range(self.num_nodes)]
        q.put((start, 1))
        while not q.empty():
            node, level = q.get()
            if not visited[node]:
                visited[node] = True
                string.append(f'->({node}, {level})')
                for edge in self.adjacency_list[node]:
                    q.put((edge.next, level + 1))
        return ''.join(map(str, string))


if __name__ == '__main__':
    num_graphs = int(input())
    for i in range(num_graphs):
        nodes, edges = map(int, input().split())
        graph = Graph(nodes)
        for j in range(edges):
            p, n = map(int, input().split('-'))
            graph.add_edge(p, n)
        print(f'-- Graph --\n{str(graph)}')
        for i in [0, 6, 3, 7]:
            print(f'-- BFS Traversal --\n{graph.bfs_traversal(i)}')
        for i in [0, 6, 3, 7]:
            print(f'-- DFS Traversal --\n{graph.dfs_traversal(i)}')

